BootloaderCorePkg/Library/S3SaveRestoreLib/S3SaveRestore.c

```diff
@@ -0,0 +1,175 @@
+/** @file
+
+  Copyright (c) 2019, Intel Corporation. All rights reserved.<BR>
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include <PiPei.h>
+#include <Library/DebugLib.h>
+#include <Library/HobLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/S3SaveRestoreLib.h>
+#include <Library/IoLib.h>
+#include <Guid/SmmInformationGuid.h>
+#include <Library/BoardInitLib.h>
+#include <Guid/SmmInformationGuid.h>
+#include <Base.h>
+
+/**
+  This function uploads information in TSEG area
+  designated for S3 save/restore purpose.
+
+  @param    DataPtr       Address of the structure to be copied to TSEG
+
+  @retval   None
+**/
+VOID
+EFIAPI
+SetS3Info (
+  IN  VOID    *DataPtr
+  )
+{
+  UINT8             *Ptr;
+  LDR_SMM_INFO      LdrSmmInfo;
+  UINT8             *SmmBase;
+  UINT32            SmmSize;
+
+  PlatformUpdateHobInfo (&gSmmInformationGuid, &LdrSmmInfo);
+  SmmBase = (UINT8 *) LdrSmmInfo.SmmBase;
+  if (LdrSmmInfo.Flags & SMM_FLAGS_4KB_COMMUNICATION) {
+    SmmSize = SIZE_4KB;
+  } else {
+    DEBUG((DEBUG_ERROR, "Invalid Communication area size\n"));
+    return;
+  }
+
+  for (Ptr = SmmBase; Ptr < SmmBase + SmmSize; ) {
+    if ( ((BL_PLD_COMM_HDR *)Ptr)->Signature == BL_PLD_COMM_SIG) {
```

Comment:
For coding, better to define a new pointer for BL_LD_COMM_HDR instead of typecasting everywhere.